### PR TITLE
Lang item cleanups

### PR DESCRIPTION
**Pull request by camsteffen**
Various cleanups related to lang items.

### CODE REVIEW

File: compiler/rustc_monomorphize/src/collector.rs [hunk 1]
```diff
@@ -1296,12 +1296,12 @@ impl<'v> RootCollector<'_, 'v> {
...
-            Err(lang_item_err) => {
+        let start_def_id = match self.tcx.lang_items().start_fn() {
+            Some(s) => s,
+            None => {
                 self.tcx
                     .sess
-                    .emit_fatal(RequiresLangItem { lang_item: lang_item_err.0.name().to_string() });
+                    .emit_fatal(RequiresLangItem { lang_item: LangItem::Start.name().to_string() });
             }
         };
```
Comment (oli-obk) [inline compiler/rustc_monomorphize/src/collector.rs:1306]: This should probably use https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.require_lang_item

**Review by oli-obk — CHANGES_REQUESTED** [review]
r=me with comments resolved

File: compiler/rustc_hir/src/lang_items.rs [hunk 2]
```diff
@@ -36,6 +36,44 @@ macro_rules! expand_group {
     };
```
Comment (bjorn3) [inline compiler/rustc_hir/src/lang_items.rs]: Typo in the commit message: s/LangaugeItems/LanguageItems

File: compiler/rustc_hir/src/lang_items.rs [hunk 3]
```diff
@@ -36,6 +36,44 @@ macro_rules! expand_group {
...
+    /// The index corresponds to the order in [`LangItem`].
+    pub items: Vec<Option<DefId>>,
+    /// Lang items that were not found during collection.
+    pub missing: Vec<LangItem>,
+    /// Mapping from [`LangItemGroup`] discriminants to all
+    /// [`DefId`]s of lang items in that group.
+    pub groups: [Vec<DefId>; NUM_GROUPS],
+}
+
+impl LanguageItems {
```
[the remark this answers is not in the document]
Comment (camsteffen) [reply compiler/rustc_hir/src/lang_items.rs:31]: Other methods are removed or changed and moved in later commits.

File: compiler/rustc_monomorphize/src/collector.rs (hunk 1 above)
Comment (camsteffen) [reply compiler/rustc_monomorphize/src/collector.rs]: done

**Comment by camsteffen** [comment]
> `rustc_error_messages` was changed

I just removed a diagnostic struct in favor of `tcx.require_lang_item`.

**Pull request merged**